cpp/src/parquet/encoding.cc

```diff
@@ -1983,7 +1990,7 @@ class DictByteArrayDecoderImpl : public DictDecoderImpl<ByteArrayType>,
     int values_decoded = 0;
 
     ArrowBinaryHelper<ByteArrayType> helper(out, num_values);
-    RETURN_NOT_OK(helper.Prepare(len_));
+    RETURN_NOT_OK(helper.Prepare());
```

Comment:
I still don't understand why we don't want to presize here.... This can be called for non-dict decoding, right?

Reply:
right!

`len_` is page size, for dict, the page size is a RLE buffer rather than a data-buffer, so this `size_` is misleading...

Reply:
Ok, so can we distinguish between dict and non-dict decoding here?

Reply:
🤔I'm not sure what do you mean `between dict and non-dict decoding`.

```c++
  inline void DecodeDict(TypedDecoder<Type>* dictionary) {
    dictionary_length_ = static_cast<int32_t>(dictionary->values_left());
    PARQUET_THROW_NOT_OK(dictionary_->Resize(dictionary_length_ * sizeof(T),
                                             /*shrink_to_fit=*/false));
    dictionary->Decode(reinterpret_cast<T*>(dictionary_->mutable_data()),
                       dictionary_length_);
  }

```

Do you mean decoding to Dict? Or just decoding a RLE_DICT encoding data? I think for decoing `RLE_DICT` or `DELTA_BYTE_ARRAY`, we cannot find the exactly length for the page, so I don't think it's proper to `Prepare` here...

Reply:
The Plain and `DeltaLengthByteArray` has a feature that it knows the remaining size in the page, so I think that's ok...

Reply:
Yes, I mean decoding to dict.

Reply:
```c++
template <>
void DictDecoderImpl<ByteArrayType>::SetDict(TypedDecoder<ByteArrayType>* dictionary) {
  DecodeDict(dictionary);

  auto dict_values = reinterpret_cast<ByteArray*>(dictionary_->mutable_data());

  int total_size = 0;
  for (int i = 0; i < dictionary_length_; ++i) {
    total_size += dict_values[i].len;
  }
  PARQUET_THROW_NOT_OK(byte_array_data_->Resize(total_size,
                                                /*shrink_to_fit=*/false));
  PARQUET_THROW_NOT_OK(
      byte_array_offsets_->Resize((dictionary_length_ + 1) * sizeof(int32_t),
                                  /*shrink_to_fit=*/false));

```

I think it already handled this well :-)

Reply:
Oops, sorry! I was misreading the context...

I think we just need an additional comment here to make sure the next reader will understand this :-)

Reply:
Maybe I missed something but I should comment on what? Decoding or other?😊

Reply:
I mean comment on why we are not passing a length here.

Reply:
done!